General | Adjust and unify password handling to increase security

MichaIng opened the pull request:
Ref: https://dietpi.com/phpbb/viewtopic.php?f=9&t=2026#p12842
- The goal is to make users aware of global plain text password and "dietpi" user.
- Further separate global password from unix user passwords for security reasons.
- Global PW (dietpi.txt) is now only used as initial password for new DietPi-Software installs and new unix users via DietPi-Set_Software. Users are asked now to set a different password for unix users, which will not saved anywhere as plain text.
+ DietPi-Globals | Add G_WHIP_PASSWORD, including double input and match check
+ DietPi-Set_Software | Add mode "passwords" with optional argument to set global password. "root" and "dietpi" login passwords can be set on a second interactive-only prompt.
+ DietPi-Software/Config | Now use new Set_Software mode instead of own password function
+ Some code improvements as always.

Testing:
- 🈯️ VM Jessie - update, config, match loop

MichaIng committed:
v6.9

+ DietPi-Set_Software | Add global password change mode, moved from DietPi-Software
+ DietPi-Set_Software | Rearrange ntpd-mode to enable and start systemd-timesyncd on mode 4 (daemon + drift)
+ DietPi-Set_Software | Minor code improvements and rearrangements, implementation of G_CONFIG_INJECT where applicable

MichaIng committed:
v6.9

+ DietPi-Software | Global password setting moved to DietPi-Set_Software

MichaIng committed:
v6.9

+ DietPi-Config | Use DietPi-Set_Software to adjust passwords

MichaIng committed:
v6.9

+ DietPi-Software | 1st run: Set chosen global password from dietpi.txt. Prompt to adjust will appear if $GLOBAL_PW == 'dietpi'. 2nd prompt will let user choose different login password for "root" and "dietpi" users to avoid having most critical passwords as plain text in dietpi.txt

MichaIng committed:
v6.9

+ DietPi-Globals | Add G_WHIP_PASSWORD for masked password inputs and secure non-global result handling

MichaIng committed:
v6.9

+ wip...

Fourdee commented:
@MichaIng 

Great work 👍 

We need to make the user aware, that changing the password after software has been installed, will not change the password for previously installed items, and, may break future installations (eg: mysql using new root password).

Ideally, the password should only be changed during 1st run, or, we have a function in ```dietpi-software``` which reapplies the global password to all software which requires one (big job).

MichaIng committed:
v6.9

+ DietPi-Set_Software | ...finished: Adjust global/default password via argument or user input and "root" + "dietpi" unix user login password separately

MichaIng committed:
v6.9

+ DietPi-Globals | G_WHIP_PASSWORD: Include doubled password entry and checking for match

MichaIng committed:
v6.9

+ DietPi-Set_Software | Mode passwords: Simplify code due to doubled input check moved to DietPi-Globals

MichaIng commented:
@Fourdee 
I went through DietPi-Software and other scripts, where we use the password scraped from dietpi.txt.
At least as far as I could see, changing it, will only change the default password for new software and users, but we never need the "old" one to automated log in somewhere. MySQL indeed needed this in the past, but with MariaDB and its unix_socket authentication "root" can log in without any password 😃.

I will test and go through the code carefully again. But for security reasons, as we take care this now from two edges (SFTP), it is good to enable users freely change all of their passwords, especially the unix user login passwords. Plain text dietpi.txt entry should be really just a 1st run, new user + software default.

MichaIng committed:
v6.9

+ DietPi-Patchfile | Offer to change global and unix user passwords to make users aware of the existance or "dietpi" and how the dietpi.txt password is used

Fourdee commented:
@MichaIng 

>  it is good to enable users freely change all of their passwords, especially the unix user login passwords. Plain text dietpi.txt entry should be really just a 1st run, new user + software default.

Yep agree, good work 👍 

If we still need access to the ```GLOBAL_PW``` in our scripts, we could prompt the user to enter it? ```dietpi.txt``` entry would then be cleared and nullified.

---

Yep, used by samba server etc, in ```dietpi-software``` we could do 
```
# at the top of Run_Installations(){
if (( $G_DIETPI_INSTALL_STAGE >= 1 )); then

please enter the root password
GLOBAL_PW=$1

fi
```

MichaIng committed:
v6.9

+ DietPi-Set_Software | Mode passwords: strings cannot be "returned" by functions, switch to variable

MichaIng committed:
v6.9

+ DietPi-Globals | G_WHIP_PASSWORD: Strings cannot be returned; Switch to safe value in variable

MichaIng committed:
v6.9

+ DietPi-Set_Software | Reduce redundant output, re-comment dbus marking, as long as we do not install libpam-systemd on desktop install/dietpi.txt option

MichaIng commented:
@Fourdee 
Work done, ready for testing.

> Yep, used by samba server etc, in dietpi-software we could do

https://github.com/Fourdee/DietPi/blob/dietpi_password/dietpi/dietpi-software#L8892
Hmm does `smbpasswd` require the current user password? Looks like it can be different than unix login passwd. https://www.samba.org/samba/docs/current/man-html/smbpasswd.8.html

I added the my ideas behind all of this above:
- Use dietpi.txt/global PW as initial password for new installed software
- But allow user do change each software passwords independently
- For this we need to assure that our scripts don't need to access the software once installed. But as mentioned, as far as I can see we just use GLOBAL_PW to set software passwords, but never to access them.

Fourdee commented:
@MichaIng 

Legend, great stuff 👍 

> Hmm does smbpasswd require the current user password? Looks like it can be different than unix login passwd.

Yeah its different. Allows users to change it if needed: https://dietpi.com/phpbb/viewtopic.php?f=8&t=5&start=10#p56

Fourdee commented:
@MichaIng 
Made a few changes, hope you don't mind 😃? :
- ```/DietPi/dietpi/func/dietpi-set_software passwords newpassword```, applies input password to both dietpi-software and root+dietpi user login. Used in automation and I want to keep this feature.
- ```/DietPi/dietpi/func/dietpi-set_software passwords``` prompts if ```G_USER_INPUTS```
- Apply passwords at bottom of script, if value set, then delete vars

Fourdee commented:
Local testing passed, great work @MichaIng 👍 

We just need to wrap up https://github.com/Fourdee/DietPi/pull/1822 tomorrow, re-test and we should be good to release tomorrow aswell. But lets see.
